test/test_torch.py

```diff
@@ -5376,6 +5376,283 @@ def test_narrow_empty(self, device):
+            test_cases.append(
+                ((4, 13, 2, 5), 'constant', pad_width, {'constant_values': constant_values}))
+
+
+        # 'constant' mode with floating point constant_values
+        if dtype.is_floating_point or dtype.is_complex:
+            test_cases += [
+                ((4, 13, 2, 5), 'constant', ((3, 2), (8, 10), (1, 0), (5, 6)),
+                    {'constant_values': ((3.3, 1.3), (4.4, 5), (10, 1), (4, 8))}),
+                ((4, 13, 2, 5), 'constant', ((3, 2), (8, 10), (1, 0), (5, 6)),
+                    {'constant_values': 5.4}),
+            ]
+
+        # 'constant' mode with complex constant_values
+        if dtype.is_complex:
+            test_cases += [
+                ((4, 13, 2, 5), 'constant', ((3, 2), (8, 10), (1, 0), (5, 6)),
+                    {'constant_values': ((3. + 1j, 1.3j), (4.4, 5), (10, 1), (4j, 8))}),
+                ((4, 13, 2, 5), 'constant', ((3, 2), (8, 10), (1, 0), (5, 6)),
+                    {'constant_values': 1.2j}),
+            ]
+
+        for input_size, mode, pad_width, kwargs in test_cases:
+            input = make_tensor(input_size, device, dtype, low=-9, high=9)
+            input_np = input.cpu().numpy()
+
+            result = input.pad(pad_width, mode, **kwargs).cpu()
+            result_np = torch.from_numpy(np.pad(input_np, pad_width, mode, **kwargs))
+
+            msg = f'input_size: {input_size}, mode: {mode}, pad_width: {pad_width}, kwargs: {kwargs}'
+
+            self.assertEqual(result.size(), result_np.size(), msg=msg)
+
+            if mode == 'empty':
+                # For empty mode, only check the non-padded elements,
+                # since the padded elements could have any value
+
+                if input.numel() == 0:
+                    continue
+
+                slices = []
+
+                pw = torch.tensor(pad_width)
+
+                if pw.numel() == 1:
+                    pw = pw.as_strided((input.dim(), 2), (0, 0))
+                elif (pw.dim() == 2) and (pw.size(0) == input.dim()) and (pw.size(1) == 1):
+                    pw = pw.as_strided((input.dim(), 2), (1, 0))
+                elif pw.numel() == 2:
+                    pw = pw.as_strided((input.dim(), 2), (0, 1))
+
+                for (before, after), dim_idx in zip(pw, range(input.dim())):
+                    slices.append([before, before + input.size(dim_idx) - 1])
+
+                self.assertEqual(result[slices], result_np[slices], msg=msg)
+
+            else:
+                self.assertEqual(result, result_np, msg=msg)
+
+    # This test ensures that torch.pad's input and constant_values arg can have
+    # different dtypes, for mode='constant'. The result should match what we
+    # get if we cast constant_values to the input's dtype before the torch.pad
+    # call.
+    @onlyOnCPUAndCUDA
+    @dtypes(*torch.testing.get_all_dtypes())
+    def test_pad_constant_values_dtypes(self, device, dtype):
```

Comment:
This is really interesting! This behavior is consistent with NumPy's pad, but @saketh-are and I were surprised that np.pad did this. That is, trying to pad an integer array by 1.5 result in padding it by 1! It seems divergent from NumPy's typical type promotion.

Should we preserve this behavior or maybe throw an error if the constant value isn't representable in the tensor's datatype, @kurtamohler?

cc @rgommers to see if he has some context on pad's (lack of) type promotion

Reply:
Just to clarify, by "pad an integer array by 1.5", I assume you mean "pad an integer array with constant padding value 1.5". Not "pad an integer array with pad width size 1.5".

I don't think this is completely unprecedented, given that a numpy array's `__setitem__` call does this:

```python
>>> a = np.arange(10)
>>> a[4] = 1.5
>>> a[4]
1
```

Reply:
> Just to clarify, by "pad an integer array by 1.5", I assume you mean "pad an integer array with constant padding value 1.5". Not "pad an integer array with pad width size 1.5".

Yes, sorry that was unclear.
 
> I don't think this is completely unprecedented, given that a numpy array's `__setitem__` call does this:
> 
> ```python
> >>> a = np.arange(10)
> >>> a[4] = 1.5
> >>> a[4]
> 1
> ```
> 
> We do the same in pytorch:
> 
> ```python
> >>> a = torch.arange(10)
> >>> a[4] = 1.5
> >>> a[4]
> tensor(1)
> ```
> 
> But I'm not against the idea of throwing an error in `torch.pad`

I don't have a great sense for which is more reasonable off-hand, either. Throwing an error always gives us more flexibility in the future, though.

Reply:
I can't remember a discussion about this, but it doesn't surprise me. Padding is conceptually different from, for example, concatenation - for padding there's already one array that you're "growing" rather than putting together multiple arrays which don't have a clear precedence.

This downcasting is perhaps something NumPy would do differently if `pad` was implemented as a new function today, we've been getting more strict about raising exceptions instead (e.g. it's no longer allowed to use floats for indexing). But upcasting everything is not expected.

Reply:
If it's easy to throw an error to avoid this issue for now that sounds best, then.

Reply:
Cool, I agree. I'll add the error

Reply:
I've added a check that throws an error if the dtype of `constant_values` is not the same as the input.

But we can also enable upcasting if that would be better. Is there a utility function somewhere that checks for upcasting cases? I haven't found one yet